File: elixir/lib/bubble_sort.ex
defmodule BubbleSort do
  def bubble_sort(list) when length(list) < 2, do: list
  def bubble_sort(list) do
    {rest, [max]} = list |> bubble_move_right |> Enum.split(-1)
    bubble_sort(rest) ++ [max]
  end

  def cocktail_shaker_sort(list) when length(list) < 2, do: list
  def cocktail_shaker_sort(list) do
    {rest, [max]} = list |> bubble_move_right |> Enum.split(-1)
    [min | rest] = rest |> bubble_move_left
    [min] ++ cocktail_shaker_sort(rest) ++ [max]
  end

  defp bubble_move_right([a]), do: [a]
  defp bubble_move_right([a, b | rest]) when a > b, do: [b | bubble_move_right([a | rest])]
  defp bubble_move_right([a, b | rest]), do: [a | bubble_move_right([b | rest])]

  defp bubble_move_left([a]), do: [a]
  defp bubble_move_left(list) do
    {rest, [a, b]} = Enum.split(list, -2)
    if a > b do
      bubble_move_left(rest ++ [b]) ++ [a]
    else
      bubble_move_left(rest ++ [a]) ++ [b]
    end
  end
end

File: ruby/lib/bubble_sort.rb
module BubbleSort
  def self.bubble_sort(orig_arr)
    arr = orig_arr.dup
    max = arr.size - 1

    while !max.zero?
      new_max = 0
      1.upto max do |i|
        if arr[i-1] > arr[i]
          arr[i-1], arr[i] = arr[i], arr[i-1]
          new_max = i-1
        end
      end
      max = new_max
    end

    arr
  end

  def self.cocktail_shaker_sort(orig_arr)
    arr = orig_arr.dup
    min, max = 1, arr.size - 1

    while min <= max
      new_max = 0
      min.upto max do |i|
        if arr[i-1] > arr[i]
          arr[i-1], arr[i] = arr[i], arr[i-1]
          new_max = i-1
        end
      end
      max = new_max
      break if max.zero?

      new_min = 0
      max.downto min do |i|
        if arr[i-1] > arr[i]
          arr[i-1], arr[i] = arr[i], arr[i-1]
          new_min = i+1
        end
      end
      min = new_min
      break if min.zero?
    end

    arr
  end
end
